src/rgw/rgw_website.cc

```diff
@@ -45,7 +45,7 @@ void RGWBWRoutingRule::apply_rule(const string& default_protocol, const string&
 
   if (!redirect_info.replace_key_prefix_with.empty()) {
     *new_url += redirect_info.replace_key_prefix_with;
-    *new_url += key.substr(condition.key_prefix_equals.size());
+    *new_url += (key.size() < condition.key_prefix_equals.size()) ? "" : key.substr(condition.key_prefix_equals.size());
```

Comment:
can you help me understand how we got to `apply_rule()` when the `key` doesn't match `condition.key_prefix_equals`? shouldn't `RGWBWRoutingRuleCondition::check_key_condition()` prevent this?

Reply:
Suppose that I have a bucket named www.example.com and my bucket contains the following objects:
index.html
error.html
docs/index.html
docs/error.html
documents/index.html
documents/error.html
I decide to redirect requests for prefix docs/ to documents/. I use the python sdk as follows:
`res = s3Client.put_bucket_website(
    Bucket='www.example.com',
    WebsiteConfiguration={
        'ErrorDocument': {
            'Key': 'error.html'
        },
        'IndexDocument': {
            'Suffix': 'index.html'
        },
        'RoutingRules': [
            {
                'Condition': {
                    'KeyPrefixEquals': 'docs/'
                },
                'Redirect': {
                    'ReplaceKeyPrefixWith': 'documents/',
                }
            },
        ]
    },
)`
When I access http://www.example.com/docs only lack the ending slash(/), the **key_name** variable is **docs** in the **RGWHandler_REST_S3Website::retarget** function. After calling **RGWBucketWebsiteConf::get_effective_key** function, the **new_obj.name** variable becomes **docs/index.html**. When calling **RGWBWRoutingRuleCondition::check_key_condition()**, the **key** variable is **docs/index.html** which can match **condition.key_prefix_equals(docs/)**. The **should_redirect** variable becomes true, then we can get to **RGWBWRoutingRule::apply_rule** function. In the **RGWBWRoutingRule::apply_rule** function, the **key** is **docs**, but the **condition.key_prefix_equals** is **docs/**. When processing **key.substr(condition.key_prefix_equals.size())**, it will cause memory read out of bounds. Thanks. @cbodley

Reply:
thanks @xrmeng8756, that makes sense. just one suggestion for the fix:
```c++
if (key.size() > condition.key_prefix_equals.size()) {
  *new_url += key.substr(condition.key_prefix_equals.size());
}
```

Reply:
I have fixed it according to your suggestion. I don't know why there are some checks fail, please help me, thanks. @cbodley